core/trino-main/src/main/java/io/trino/operator/MergeHashSort.java

```diff
@@ -73,11 +73,20 @@ public void close()
 
     private static BiPredicate<PageBuilder, PageWithPosition> keepSameHashValuesWithinSinglePage(InterpretedHashGenerator hashGenerator)
     {
-        return (pageBuilder, pageWithPosition) -> {
-            long hash = hashGenerator.hashPosition(pageWithPosition.getPosition(), pageWithPosition.getPage());
-            return !pageBuilder.isEmpty()
-                    && hashGenerator.hashPosition(pageBuilder.getPositionCount() - 1, pageBuilder::getBlockBuilder) != hash
-                    && pageBuilder.isFull();
+        return new BiPredicate<>()
+        {
+            private long lastHash;
+
+            @Override
+            public boolean test(PageBuilder pageBuilder, PageWithPosition pageWithPosition)
```

Comment:
If I understand correctly, the old implementation checked whether the given position has the same hash as the last position in the provided page builder.

The new implementation does something very different. In particular, it doesn't even look at the pageBuilder, and only compares hashes across subsequent invocations. How are they equivalent?

For instance, given a pageBuilder last position hash = 10 and the following input hashes, the results would differ:

|position hash|old|new|
|-|-|-|
|1|F|F|
|1|F|T|
|10|T|F|
|3|F|F|

I guess I don't quite understand the purpose of this method (`keepSameHashValuesWithinSinglePage`).

Is the pageBuilder changing on every invocation in the previous implementation such that the last position is the previous position tested by this method?

Reply:
Yes the page builder is the same.  The goals of this code is recorded in the class Javadocs

> Rows with same hash value are guaranteed to be in the same result page.

I'm pretty sure this code returns the same results as the existing code, with the asumption the page builder is the same.  I think you mised the `!sameHash` in the return statement